py/server/deephaven/table.py

```diff
@@ -363,108 +368,306 @@ def _j_py_script_session() -> _JPythonScriptSession:
-    Deephaven primitive null values.
-
-    For properly annotated functions, including numba vectorized and guvectorized ones, this decorator inspects the
-    signature of the function and determines its return type, including supported primitive types and arrays of
-    the supported primitive types. It then converts the return value of the function to the corresponding Java value
-    of the same type. For unsupported types, the decorator returns the original Python value which appears as
-    org.jpy.PyObject in Java.
+    It carries out two conversions:
+    1. convert Python function return values to Java values.
+        For properly annotated functions, including numba vectorized and guvectorized ones, this decorator inspects the
+        signature of the function and determines its return type, including supported primitive types and arrays of
+        the supported primitive types. It then converts the return value of the function to the corresponding Java value
+        of the same type. For unsupported types, the decorator returns the original Python value which appears as
+        org.jpy.PyObject in Java.
+    2. convert Java function arguments to Python values based on the signature of the function.
     """
-
     if hasattr(fn, "return_type"):
         return fn
-    ret_dtype = _udf_return_dtype(fn)
-
-    return_array = False
-    # If the function is a numba guvectorized function, examine the signature of the function to determine if it
-    # returns an array.
-    if isinstance(fn, numba.np.ufunc.gufunc.GUFunc):
-        sig = fn.signature
-        rtype = sig.split("->")[-1].strip("()")
-        if rtype:
-            return_array = True
-    else:
-        try:
-            return_annotation = _parse_annotation(inspect.signature(fn).return_annotation)
-        except ValueError:
-            # the function has no return annotation, and since we can't know what the exact type is, the return type
-            # defaults to the generic object type therefore it is not an array of a specific type,
-            # but see (https://github.com/deephaven/deephaven-core/issues/4762) for future imporvement to better support
-            # numpy ufuncs.
-            pass
-        else:
-            component_type = _component_np_dtype_char(return_annotation)
-            if component_type:
-                ret_dtype = dtypes.from_np_dtype(np.dtype(component_type))
-                if ret_dtype in _BUILDABLE_ARRAY_DTYPE_MAP:
-                    return_array = True
+    p_sig = _parse_signature(fn)
+    return_array = p_sig.ret_annotation.is_array
+    ret_dtype = dtypes.from_np_dtype(np.dtype(list(p_sig.ret_annotation.encoded_types)[0][-1]))
 
     @wraps(fn)
     def wrapper(*args, **kwargs):
-        ret = fn(*args, **kwargs)
+        converted_args = _convert_args(p_sig, args)
+        ret = fn(*converted_args, **kwargs)
```

Comment:
I would add a comment about why `kwargs` are not converted.  Potentially create a ticket to handle conversion after we support `kwargs` in the query lang.